### ISSUE TITLE
Only send needed files as Docker context (use Buildkit)

### ISSUE THREAD

**Issue by djbrown** [openshift/imagebuilder#143]
Today I searched for a method to enhance the performance for a Docker Build in OpenShift by only sending those files as Docker context, which we actually copy/add/use in the Dockerfile.
Using a .dockerignore file was no solution, since we use multiple repositories and I would have needed multiple compound ignore files.

Then I found out about **[Buildkit](https://github.com/moby/buildkit)** and how it only sends those files which are used in the Dockerfile - exactly what I wanted.

None of the OpenShift docs mention Buildkit ([search results](https://docs.openshift.com/search.html?query=buildkit)).
Then I searched in the openshift organsization on GitHub, but no relevant [Code results](https://github.com/search?p=1&q=org%3Aopenshift+buildkit&type=Code) came up.
The only relevant result was https://github.com/openshift/imagebuilder/pull/90, which was closed.
_Originally posted by @rhatdan in https://github.com/openshift/imagebuilder/pull/90#issuecomment-407511847_
> Since this did not actually fix the issue, and the issue I saw failed the same way on Docker, I am closing this.

I found a commit similar to the preceding PR https://github.com/openshift/imagebuilder/commit/67132c9404d6510931c43a1fbc9e1270875b29db, which does not seem to be in either of the current branches.

# How can I use Buildkit for Docker BuildConfigs?

**Comment by djbrown** [comment]
can a maintainer pls take a look at this?

**Comment by djbrown** [comment]
@rhatdan @TomSweeneyRedHat

**Comment by nalind** [comment]
imagebuilder uploads what it needs to in order to handle ADD and COPY instructions as it goes along.
```
mkdir tmp
dd if=/dev/urandom of=tmp/bigfile bs=1M count=10K
touch tmp/smallfile
cat > tmp/Dockerfile << EOF
FROM busybox
ADD smallfile /
EOF
imagebuilder ./tmp
docker build ./tmp
```
We ended up forking the Dockerfile parser around #125 to avoid pulling in other dependencies.

**Comment by TomSweeneyRedHat** [comment]
@djbrown does @nalind's comment answer your question?  If so, can you close this issue?

**Comment by djbrown** [comment]
Not really, but when digging deeper to explain my problem I found a workaround/solution.
Yet this issue might still be eligible.

My original concern was build perfomance when using binary remote build.
I was using the OCP server as "docker service/host", sending my local files as docker context via `oc start-build --from-dir="."`.
This would upload the whole directory which took a lot of time.

In the docs under [Build Inputs > Binary (Local) Source](https://docs.openshift.com/container-platform/3.11/dev_guide/builds/build_inputs.html#binary-source) I found the `--from-archive` option.
This way I can improve the performance *manually*.

Ideally the oc command would analyze the Dockerfile definition beforehand and only upload needed files, but I think the command wasn't design for that. Is this something you would consider  implementing?
If this is the wrong component/repo for this issue I will move it.
If this is an  overall unreasonable request (by design) then you or I may close this issue.

P.S. Also I discovered that in https://github.com/moby/moby/commit/c9ebd2f13b3aab49c2d90cc38dcf2cd0d059697f `builder/dockerfile/parser/` was actually removed/replaced by `moby/buildkit/frontend/dockerfile/parser/`. Maybe BuildKit is more isolated and could be included with less unused dependencies.
